docs/dev/src/design/keys.md

```diff
@@ -23,15 +42,39 @@ The record corresponding to key `(0, 1)` has been inserted with `(0, 1, 2, 3)`.
 It may not be the minimal set of columns required to identify a record,
 for instance `group key` could be part of the stream key, to specify the distribution of records.
 
-## Primary Key (Storage)
+## Distribution key
+
+This key specifies the distribution of the data streams, as mentioned in [Consistent Hash - Streaming](../design/consistent-hash.md#streaming).
+We want data to be distributed in a way that minimizes data skew,
+and maximizes data locality, for more efficient stateful processing.
+
+To ensure data consistency of updates (U-, U+),
+the distribution key must always be a prefix of the stream key.
+See [issue #21668](https://github.com/risingwavelabs/risingwave/issues/21668) for an example.
+This ensures that updates on the same key are not sent to different partitions.
+Otherwise if they are exchanged back to the same partition,
+the order of updates may be different from the order of the original stream.
+
+## Primary Key
+
+### User Specified Primary Key
+
+When the user specifies primary key in their `create table` statement,
+it is used to ensure that the records are unique in the table.
+We will always use the user-specified primary key as the stream key and storage primary key
+
+For other streaming jobs such as sinks, materialized views and indexes,
+we will use the user-specified primary key in our derivation of the stream key and storage primary key.
```

Comment:
ditto

Reply:
I think to understand storage primary key, need to first understand what "Table" means.. (quite messy)

A table = TableCatalog = a hummock table = internal state table/ MV (i.e., the state table for MaterializeExec) / ...